### PR TITLE
Adding a proposal for handling metrics.

### PR DESCRIPTION
**Pull request by vishh**
I am posting a WIP to gain some initial feedback on the overall design. 
This is critical for designing auto-scalers and improving the current scheduler.

Kindly do not nit-pick just yet. Some of the links might be broken.

cc @piosz @fgrzadkowski @mwielgus @smarterclayton @bgrant0607 @lavalamp

### CODE REVIEW

File: docs/proposals/metrics-plumbing.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+There are a variety of storage backends that can handle metrics.
+
+This document presents a design to abstract out collection and storage backends, and provide stable Kubernetes APIs that can be consumed by users and other cluster components.
+
+## Introduction
+
+Container metrics can be of two types.
+
+1. `Compute resource metrics` refers to compute resources being used by a container. Ex.: CPU, Memory, Network, File-system
+2. `Service metrics` refers to container app specific metrics. Ex: QPS, query latency, etc.
```
Comment (smarterclayton) [inline docs/proposals/metrics-plumbing.md]: I assume you mean service in the generic sense - logical groups of pods or external endpoints with a given name.

File: docs/proposals/metrics-plumbing.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+Metrics can be collected either for cluster components or for user containers.
+
+[cAdvisor](https://github.com/google/cadvisor) is a node level container metrics aggregator that is built into the kubelet. cAdvisor can collect both types of metrics, although the support for service metrics is limited at this point. cAdvisor collects metrics for both system components and user containers.
+
+[heapster](https://github.com/kubernetes/heapster) is a cluster level metrics aggregator that is run by default on most Kubernetes cluster. Heapster aggregates all the metrics exposed by cAdvisor from the nodes. Heapster has a pluggable storage backend. It supports the following timeseries storage backends - InfluxDB, Google Cloud Monitoring and Hawkular.
+Heapster builds a model of the cluster and can aggregate metrics across pods, nodes, namespaces and the entire cluster. It exposes this data via [REST endpoints](https://github.com/kubernetes/heapster/blob/master/docs/model.md#api-documentation).
+
+Metrics data will be consumed by many different clients - scheduler, horizontal and vertical pod auto scalers, initial pod limits controller, kubectl, etc.
```
Comment (smarterclayton) [inline docs/proposals/metrics-plumbing.md]: web consoles, cluster management software

**Comment by smarterclayton** [comment]
Pretty good

**Comment by brendandburns** [comment]
closing and re-opening to trigger travis/shippable

**Comment by piosz** [comment]
cc @jszczepkowski

File: docs/proposals/metrics-plumbing.md (hunk 1 above)
Comment (vishh) [inline docs/proposals/metrics-plumbing.md]: I meant app metrics like nginx queries per second, response latency, etc.

Comment (jimmidyson) [inline docs/proposals/metrics-plumbing.md]: How are app metrics exposed to heapster?

Comment (vishh) [inline docs/proposals/metrics-plumbing.md]: [cAdvisor](https://github.com/google/cadvisor/tree/master/collector) to begin with. We can extend support for collectd or other collectors if necessary, but we need to somehow whitelist the metrics that matter to the rest of the system (auto-scaling mainly).

**Comment by vishh** [comment]
@davidopp

File: docs/proposals/metrics-plumbing.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+* Requiring users to run a specific storage backend.
+* Compatibility with other node level metrics aggregator. cAdvisor should be able to provide all the metrics.
+
+## Design
+
+The basic idea is to evolve heapster to serve Metrics APIs which can then be consumed by other cluster components. 
+Heapster will be run in all clusters by default. Heapster's memory usage is proportional to the number of containers in the cluster and so it should be possible to run heapster by default even on small development or test clusters.
+A cluster administrator will have to either run one of the supported storage backends or write a new storage plugin in heapster to support custom storage backends.
+Heapster will manage versioning and storage schema for the various storage backends it supports.
+Heapster APIs will be exposed as Kubernetes APIs once the apiserver supports [dynamic API plugins](#991).
```
Comment (dchen1107) [inline docs/proposals/metrics-plumbing.md]: The link to dynamic API plugins is wrong here.

File: docs/proposals/metrics-plumbing.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+Heapster stores a days worth of historical metrics. Heapster will fetch data from storage backends on-demand to serve metrics that are older than a day. Setting [initial pod resources](#12149) requires access to metrics from the past 30 days.
```
Comment (dchen1107) [inline docs/proposals/metrics-plumbing.md]: The link to initial pod resources is wrong too.

File: docs/proposals/metrics-plumbing.md [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+Some of the specific work items include,
+
+1. Improve the existing API schema to be Kubernetes compatible ([Related issue](https://github.com/kubernetes/heapster/issues/476))
+2. Add support for fetching historical data from storage backends.
+3. Fetch historical metrics from storage backends upon restarts to pre-populate the internal model.
+4. Add support for image based aggregation.
+5. Add support for label queries.
+6. Expose heapster APIs via a Kubernetes service until the primary API server can handle plugins.
+
+Support for service metrics at the cluster level is out of scope for this document. Once the use cases for service metrics, other than monitoring, are clear, we can explore adding support for service metrics.
```
Comment (dchen1107) [inline docs/proposals/metrics-plumbing.md]: Can you move this paragraph to non-goal?

**Comment by dchen1107** [comment]
Beside several nits listed above, LGTM in content.

File: docs/proposals/metrics-plumbing.md (hunk 2 above)
Comment (vishh) [inline docs/proposals/metrics-plumbing.md]: Added.

File: docs/proposals/metrics-plumbing.md (hunk 5 above)
Comment (vishh) [inline docs/proposals/metrics-plumbing.md]: Done

File: docs/proposals/metrics-plumbing.md (hunk 4 above)
Comment (vishh) [inline docs/proposals/metrics-plumbing.md]: Fixed

File: docs/proposals/metrics-plumbing.md (hunk 3 above)
Comment (vishh) [inline docs/proposals/metrics-plumbing.md]: Fixed.

**Comment by vishh** [comment]
@dchen1107: Fixed the issues. PTAL.

File: docs/proposals/metrics-plumbing.md [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+Some of the specific work items include,
+
+1. Improve the existing API schema to be Kubernetes compatible ([Related issue](https://github.com/kubernetes/heapster/issues/476))
+2. Add support for fetching historical data from storage backends.
+3. Fetch historical metrics from storage backends upon restarts to pre-populate the internal model.
+4. Add support for image based aggregation.
+5. Add support for label queries.
+6. Expose heapster APIs via a Kubernetes service until the primary API server can handle plugins.
+
+### Non Goals
```
Comment (dchen1107) [inline docs/proposals/metrics-plumbing.md]: You already have a non goal at line 77 above, can we combine them together?

**Comment by jimmidyson** [comment]
I've recently been working a lot with [Prometheus](http://prometheus.io) & was just wondering if anyone thought it had a part to play in the Heapster story? A lot of the goals I read in this doc can be achieved with Prometheus & I could see Heapster being a layer above Prometheus to provide a Kubernetes style API to the flexible data model that Prometheus provides. Prometheus now has autodiscovery for Kubernetes cluster components, which are instrumented already (e.g. API server requests, SkyDNS lookups, etc). It also can ingest application level metrics, discovered via annotations on services & labelled with Kubernetes labels. There is also pretty wide support for Prometheus exporters for various languages, frameworks & software.

Heapster could then provide an API above Prometheus to understand the raw Prometheus data model into in Kubernetes terms.

**Comment by vishh** [comment]
@jimmidyson: That makes sense. The storage layer can be Prometheus. We need a consistent schema though.

**Comment by jimmidyson** [comment]
So there's no schema per se in Prometheus but we can provide that via the heapster (kubernetes metrics?) api.

A metrics api in kubernetes api server would allow for potentially multiple implementations.

I'd love to see stuff like `kubectl metrics <pod>`.

File: docs/proposals/metrics-plumbing.md (hunk 6 above)
Comment (vishh) [inline docs/proposals/metrics-plumbing.md]: Yeah. Makes sense. Fixed.

**Comment by vishh** [comment]
@jimmidyson: @mvdan has a working prototype for node metrics. Adding pod metrics should be straight-forward. Let's sync up offline (on IRC), and I can provide you more information/pointers.

**Comment by dchen1107** [comment]
ok to test

**Comment by dchen1107** [comment]
LGTM

**Pull request merged**